File: scripts/action_client_node.py
#!/usr/bin/env python

"""
.. module:: action_client_node
   :platform: Unix
   :synopsis: Python ROS node for controlling robot navigation via actionlib.

.. moduleauthor:: Rubin Khadka Chhetri

This module implements an **action client** for the ROS navigation stack. It allows users to:
    - Send **target coordinates** (goals) to the robot.
    - Cancel ongoing goals.
    - Monitor robot status (position/velocity) via `/odom`.
    - Publish robot status to `/robot_status`.

Subscribes to:
    - **/odom** (`nav_msgs/Odometry`): Current robot position and velocity.

Publishes to:
    - **/robot_status** (`assignment2_rt_part1/robot_status`): Custom message with robot state.

Actions:
    - **/reaching_goal** (`assignment_2_2024/PlanningAction`): Action server for navigation.

Dependencies:
    - ``rospy``
    - ``actionlib``
    - ``assignment_2_2024``
    - ``assignment2_rt_part1``
"""

import rospy
import actionlib
from nav_msgs.msg import Odometry
from assignment_2_2024.msg import PlanningAction, PlanningGoal, PlanningFeedback
from assignment2_rt_part1.msg import robot_status

# Global variables
current_x = 0.0
"""float: Current x-coordinate of the robot (meters)."""
current_y = 0.0
"""float: Current y-coordinate of the robot (meters)."""
vel_x = 0.0
"""float: Linear velocity in x-direction (m/s)."""
vel_z = 0.0
"""float: Angular velocity around z-axis (rad/s)."""
cancel_goal_flag = False
"""bool: Flag to request goal cancellation."""
goal_active = False
"""bool: Flag indicating if a goal is active."""

def odom_callback(msg):
    """
    Callback for `/odom` subscriber. Updates robot pose and velocity.

    Args:
        msg (Odometry): Incoming message with robot state.
    """
    global current_x, current_y, vel_x, vel_z
    current_x = msg.pose.pose.position.x
    current_y = msg.pose.pose.position.y
    vel_x = msg.twist.twist.linear.x
    vel_z = msg.twist.twist.angular.z

def feedback_callback(feedback):
    """
    Handles feedback from the action server.

    Args:
        feedback (PlanningFeedback): Feedback message containing status.
    """
    if feedback.stat == "Target reached!":
        rospy.loginfo("Goal reached! Press 'Enter' to continue.")

def send_goal(client, target_x, target_y):
    """
    Sends a navigation goal to the action server.

    Args:
        client (SimpleActionClient): Action client connected to `/reaching_goal`.
        target_x (float): Target x-coordinate (meters).
        target_y (float): Target y-coordinate (meters).
    """
    global goal_active
    goal = PlanningGoal()
    goal.target_pose.pose.position.x = target_x
    goal.target_pose.pose.position.y = target_y
    goal.target_pose.pose.orientation.w = 1.0  # Neutral orientation
    rospy.loginfo(f"Sending goal: ({target_x}, {target_y})")
    client.send_goal(goal, feedback_cb=feedback_callback)
    goal_active = True

def cancel_goal(client):
    """
    Cancels the current goal if requested.

    Args:
        client (SimpleActionClient): Action client to send cancel request.
    """
    global cancel_goal_flag, goal_active
    if cancel_goal_flag and goal_active:
        client.cancel_goal()
        rospy.loginfo("Goal cancellation requested.")

def action_client_node():
    """
    Main function for the action client node.

    Handles:
    - User input for goals.
    - Goal cancellation.
    - Robot status publishing.
    """
    global cancel_goal_flag, goal_active

    rospy.init_node('action_client_node')
    client = actionlib.SimpleActionClient('/reaching_goal', PlanningAction)
    client.wait_for_server()
    rospy.loginfo("Action server connected.")

    rospy.Subscriber('/odom', Odometry, odom_callback)
    pub = rospy.Publisher('/robot_status', robot_status, queue_size=10)

    while not rospy.is_shutdown():
        try:
            # Get user input
            target_x = float(input("Enter target x: "))
            target_y = float(input("Enter target y: "))
            send_goal(client, target_x, target_y)

            # Monitor goal progress
            while not rospy.is_shutdown() and client.get_state() not in [
                actionlib.GoalStatus.SUCCEEDED,
                actionlib.GoalStatus.ABORTED,
                actionlib.GoalStatus.REJECTED
            ]:
                user_input = input("Enter 'cancel' to abort: ").strip().lower()
                if user_input == 'cancel':
                    cancel_goal_flag = True
                    cancel_goal(client)
                    break

                # Publish status
                msg = robot_status()
                msg.x = current_x
                msg.y = current_y
                msg.vel_x = vel_x
                msg.vel_z = vel_z
                pub.publish(msg)
                rospy.sleep(0.5)

            # Post-goal handling
            if client.get_state() == actionlib.GoalStatus.SUCCEEDED:
                rospy.loginfo("Ready for new coordinates.")
            elif cancel_goal_flag:
                rospy.loginfo("Goal canceled.")
            else:
                rospy.logwarn(f"Goal failed: {client.get_state()}")

            cancel_goal_flag = goal_active = False

        except ValueError:
            rospy.logerr("Invalid input! Use numbers only.")

if __name__ == '__main__':
    try:
        action_client_node()
    except rospy.ROSInterruptException:
        rospy.loginfo("Node terminated.")

File: src/action_client_node.cpp
#include <ros/ros.h>
#include <actionlib/client/simple_action_client.h>
#include <assignment_2_2024/PlanningAction.h>
#include <assignment2_rt_part1/robot_status.h>
#include <nav_msgs/Odometry.h>
#include <sstream>
#include <iostream>
#include <limits>

// Global variables to store current position and velocity
double current_x = 0.0, current_y = 0.0, vel_x = 0.0, vel_z = 0.0;
bool cancel_goal_flag = false; // Flag to handle cancelation
bool goal_active = false;

// Function declarations
void odomCallback(const nav_msgs::Odometry::ConstPtr& msg);
bool getValidCoordinate(const std::string& prompt, double& coordinate);
void sendGoal(actionlib::SimpleActionClient<assignment_2_2024::PlanningAction>& ac, double target_x, double target_y);
void cancelGoal(actionlib::SimpleActionClient<assignment_2_2024::PlanningAction>& ac);

// Main function
int main(int argc, char** argv) {
    // Initialize ROS
    ros::init(argc, argv, "action_client_node");
    ros::NodeHandle nh;

    // Set up action client
    actionlib::SimpleActionClient<assignment_2_2024::PlanningAction> ac("/reaching_goal", true);
    ac.waitForServer();
    ROS_INFO("Action Server Started !!!");

    // Subscribe to /odom to get position and velocity
    ros::Subscriber sub_odom = nh.subscribe("/odom", 10, odomCallback);

    // Publisher for position and velocity
    ros::Publisher pub_pos_vel = nh.advertise<assignment2_rt_part1::robot_status>("/robot_status", 10);

    while (ros::ok()) {
        try {
            double target_x, target_y;

            // Get valid user coordinates for the goal
            if (!getValidCoordinate("Enter target x:", target_x)) {
                continue;
            }

            if (!getValidCoordinate("Enter target y:", target_y)) {
                continue;
            }

            // Send the goal to the action server
            sendGoal(ac, target_x, target_y);

            // Monitor the goal state and robot's position while the goal is in progress
            while (!ac.getState().isDone() && ros::ok()) {
                // Check for cancel command from user input
                std::string user_input;
                std::cout << "Robot Moving!!! Enter 'cancel' to cancel the goal: ";
                std::cin >> user_input;

                if (user_input == "cancel") {
                    cancel_goal_flag = true;
                    cancelGoal(ac);
                    std::cin.clear(); // Clear error
                    std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
                    break;
                }

                // Publish robot's current position and velocity
                assignment2_rt_part1::robot_status pos_vel_msg;
                pos_vel_msg.x = current_x;
                pos_vel_msg.y = current_y;
                pos_vel_msg.vel_x = vel_x;
                pos_vel_msg.vel_z = vel_z;
                pub_pos_vel.publish(pos_vel_msg);

                ros::Duration(0.5).sleep();  // Sleep for 500 ms to avoid spamming messages
            }

            // Check the outcome of the goal
            if (ac.getState() == actionlib::SimpleClientGoalState::SUCCEEDED) {
                ROS_INFO("Goal reached successfully!");
            } else if (cancel_goal_flag) {
                ROS_INFO("Goal was canceled by the user.");
            } else {
                ROS_INFO("Goal did not succeed. Status: %d", ac.getState().state_);
            }

            cancel_goal_flag = false;  // Reset cancel flag
            goal_active = false;

        } catch (const std::invalid_argument& e) {
        }

        ros::Duration(1).sleep();  // Sleep for 1 second before accepting new input
    }

    ros::shutdown();
    return 0;
}

// Callback to update the robot's position and velocity from /odom
void odomCallback(const nav_msgs::Odometry::ConstPtr& msg) {
    current_x = msg->pose.pose.position.x;
    current_y = msg->pose.pose.position.y;
    vel_x = msg->twist.twist.linear.x;
    vel_z = msg->twist.twist.angular.z;
}

// Function to get valid user input for coordinates
bool getValidCoordinate(const std::string& prompt, double& coordinate) {
    std::string input;
    while (true) {
        std::cout << prompt << " ";
        std::getline(std::cin, input);

        try {
            size_t pos;
            coordinate = std::stod(input, &pos);  // Try converting to double
            
            // Check if the whole string was consumed by std::stod
            if (pos == input.length()) {
                return true;  // Valid number, no leftover characters
            } else {
                std::cerr << "Invalid input! Please enter a valid number." << std::endl;
            }
        } catch (const std::invalid_argument&) {
            std::cerr << "Invalid input! Please enter a valid number." << std::endl;
        }
    }
}

void sendGoal(actionlib::SimpleActionClient<assignment_2_2024::PlanningAction>& ac, double target_x, double target_y) {
    assignment_2_2024::PlanningGoal goal;  // Create a goal instance
    goal.target_pose.pose.position.x = target_x;  // Set the target x-coordinate
    goal.target_pose.pose.position.y = target_y;  // Set the target y-coordinate
    goal.target_pose.pose.position.z = 0.0;
    goal.target_pose.pose.orientation.w = 1.0;

    // Log the goal
    ROS_INFO("Sending goal: x=%.2f, y=%.2f", target_x, target_y);

    // Set the target coordinates as ROS parameters
    ros::param::set("last_target_x", target_x);
    ros::param::set("last_target_y", target_y);

    // Send the goal to the action server and listen for feedback
    ac.sendGoal(goal);  // Set up feedback callback
    goal_active = true;  // Mark the goal as active
}

void cancelGoal(actionlib::SimpleActionClient<assignment_2_2024::PlanningAction>& ac) {
    if (cancel_goal_flag && goal_active) {
        // Cancel the active goal if the cancel flag is set
        ac.cancelGoal();
    }
}